backend/editor/entries.py

```diff
@@ -78,34 +82,63 @@ def import_from_github(self, description):
         filename = self.taxonomy_name + '.txt'
         base_url += filename
         try:
-            # Downloads and creates taxonomy file in current working directory
-            urllib.request.urlretrieve(base_url, filename)
-            status = self.parse_taxonomy(filename) # Parse the taxonomy
-            self.create_project(description) # Creates a "project node" in neo4j
-            os.remove(filename) # Removes taxonomy file after parsing
-            return status
+            with tempfile.TemporaryDirectory(prefix="taxonomy-") as tmpdir:
+
+                # File to save the downloaded taxonomy
+                filepath = f"{tmpdir}/{filename}"
+
+                # Downloads and creates taxonomy file in current working directory
+                urllib.request.urlretrieve(base_url, filepath)
+
+                status = self.parse_taxonomy(filepath) # Parse the taxonomy
+                self.create_project(description) # Creates a "project node" in neo4j
+                return status
         except:
-            return TaxnonomyImportError()
+            raise TaxnonomyImportError()
     
-    def export_taxonomy(self):
+    def export_taxonomy(self, type_of_export):
         """
-        Helper function to export a taxonomy for download
+        Helper function to export a taxonomy through download or Github
         """
         # Close current transaction to use the session variable in unparser
         get_current_transaction().commit()
 
+        self.close_project() # Close the project
+
         # Create unparser object and pass current session to it
         unparser_object = unparser.WriteTaxonomy(get_current_session())
         filename = self.taxonomy_name + '.txt'
         try:
             # Parse taxonomy with given file name and branch name
             unparser_object(filename, self.branch_name, self.taxonomy_name)
-            return filename
+            if type_of_export == "download":
+                return filename
+            elif type_of_export == "github":
+                return self.export_to_github(filename)

```

Comment:
(Here I'm really picky, but I know you like to learn :-) )

When you add a parameter and the return type is different based on this, this is a good example of a "code smell", that means it is not the best way to factor the code.

To make it better, I would have two different function but calling a common dump_taxonomy function (containing the above code, but the close project part)
```python
    def dump_taxonomy():
         # Create unparser object and pass current session to it
        unparser_object = unparser.WriteTaxonomy(get_current_session())
        filename = self.taxonomy_name + '.txt'   
        try:
            # Parse taxonomy with given file name and branch name
            unparser_object(filename, self.branch_name, self.taxonomy_name)
            return filename
        except Exception:
             raise TaxonomyUnparsingError()

    def file_export(self):
        # Close current transaction to use the session variable in unparser
        get_current_transaction().commit()
        filepath = self.dump_taxonomy()
        return filepath

    def github_export(self):
        # Close current transaction to use the session variable in unparser
        get_current_transaction().commit()
        filepath = self.dump_taxonomy()
        result = self.export_to_github(filepath)
        self.close_project()
        return result
```

You might think it's a bit of repetition, but I think this kind of high level repetition do not harm, as really the logic tend to be a bit different between both cases (and might differ more)

Reply:
That's really informative. Thanks for the tip!